pilot/pkg/model/jwks_resolver.go

```diff
@@ -384,41 +370,71 @@ func (r *JwksResolver) refresher() {
 	for {
 		select {
 		case <-r.refreshTicker.C:
-			currentHasError := r.refresh()
-			if currentHasError {
-				if lastHasError {
-					// update to exponential backoff if last time also failed.
-					r.refreshInterval *= 2
-					if r.refreshInterval > JwtPubKeyRefreshIntervalOnFailureResetThreshold {
-						r.refreshInterval = JwtPubKeyRefreshIntervalOnFailureResetThreshold
+			if !stopBackgroundRefresh {
+				currentHasError := r.refresh(networkFetchRetryCountOnRefreshFlow)
+				if currentHasError {
+					if lastHasError {
+						// update to exponential backoff if last time also failed.
+						r.refreshInterval *= 2
+						if r.refreshInterval > JwtPubKeyRefreshIntervalOnFailureResetThreshold {
+							r.refreshInterval = JwtPubKeyRefreshIntervalOnFailureResetThreshold
+						}
+					} else {
+						// change to the refreshIntervalOnFailure if failed for the first time.
+						r.refreshInterval = r.refreshIntervalOnFailure
 					}
 				} else {
-					// change to the refreshIntervalOnFailure if failed for the first time.
-					r.refreshInterval = r.refreshIntervalOnFailure
+					// reset the refresh interval if success.
+					r.refreshInterval = r.refreshDefaultInterval
 				}
-			} else {
-				// reset the refresh interval if success.
-				r.refreshInterval = r.refreshDefaultInterval
+				lastHasError = currentHasError
+				r.refreshTicker.Reset(r.refreshInterval)
 			}
-			lastHasError = currentHasError
-			r.refreshTicker.Reset(r.refreshInterval)
 		case <-closeChan:
 			r.refreshTicker.Stop()
 			return
+		case jwtKeyData := <-jwksuriChannel:
+			var jwksURI string
+			var err error
+			now := time.Now()
+			stopBackgroundRefresh = true
+			if jwtKeyData.jwksURI == "" {
+				// Fetch the jwks URI if it is not hardcoded on config.
+				jwksURI, err = r.resolveJwksURIUsingOpenID(jwtKeyData.issuer)
+				if err != nil {
+					log.Errorf("Failed to jwks URI from %q: %v", jwtKeyData.issuer, err)
+				}
+				jwtKeyData.jwksURI = jwksURI
+			}
+			r.keyEntries.Store(jwtKeyData, jwtPubKeyEntry{
+				pubKey:            "",
+				lastRefreshedTime: now,
+				lastUsedTime:      now,
+			})
+			hasErrors := true
+			for i := 0; i < 10; i++ {
+				if !r.refresh(0) {
+					hasErrors = false
+					break
+				}
+			}
+			if hasErrors {
+				log.Errorf("Failed to get public key from  %q: %q", jwtKeyData.issuer, jwtKeyData.jwksURI)
+			}
+			stopBackgroundRefresh = false
 		}
 	}
 }
 
-func (r *JwksResolver) refresh() bool {
+func (r *JwksResolver) refresh(retryNumber int) bool {
```

Comment:
If you reuse the refresh logic (see comment above), you shouldn't need the `retryNumber` argument.